### PR TITLE
Bench: PMem Allocator performance

### PR DESCRIPTION
**Pull request by sherry-1001**
### What problem does this PR solve?
1. Refactoring test cases
2.  Add pmem allocator tests: correctness and performance.
     a. Performance:
          Random free and allocated ops ( a fixed num ops/thread). Count allocated time and free time to compared pmem allocator performance and standard allocator(malloc).
     b. Correctness:
         Case one:
           Using different config params( `num_segment_blocks`, `block_sizes`, `num_thread`) to check pmem allocate and free: 
                  After all pmem space allocated,  free all pmem space, and again allocated. 
         Case  two:
              Different pmem_sizes to check `pmem populate space`
         Case Three:
              Memory fragemention case to check pmem memory merge.
Tests 
- Unit test
- Integration test
- No code

Side effects
    No

### CODE REVIEW

File: tests/allocator.hpp [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    mem_op.entry = pmem_alloc_->Allocate(alloc_size);
+    END
+  }
+
+  op_alloc_info wrapped_free(op_alloc_info *data) {
+    std::chrono::system_clock::time_point to_begin =
+        std::chrono::high_resolution_clock::now();
+    pmem_alloc_->Free(data->entry);
+    std::chrono::system_clock::time_point to_end =
+        std::chrono::high_resolution_clock::now();
```
Comment (ZiyanShi) [inline tests/allocator.hpp:111]: `std::chrono::high_resolution_clock::now()` has an overhead of about 20ns, an unneglectable amount compared to simple operations such as malloc and free, which usually takes several nanoseconds.
Consider do millions of operations and then measure the duration.

File: tests/test_pmem_allocator.cpp [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+    PrintRecord("pmem spend_on_alloc(ops/ms)",
+                pmem_time.allocated_ops_per_seconds);
+    PrintRecord("pmem spend_on_free(ops/ms)",
+                pmem_time.deallocated_ops_per_seconds);
+    PrintRecord("pmem spend_on_merge_time", pmem_time.pmem_merge_time);
+    float alloc_ref_delta_time_percent =
+        ((pmem_time.alloc_total_time / ref_time.alloc_total_time) - 1.0) *
+        100.0;
+    delete pmem_alloc;
+  }
```
Comment (ZiyanShi) [inline tests/test_pmem_allocator.cpp:152]: Test is test. Performance measurement should be put in benchmark.

File: tests/test_pmem_allocator.cpp [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+    std::vector<int> calls;
+    for (int i = 0; i < call_num; i++) {
+      calls.push_back(rand() % NUM_FUNCS);
+    }
+    return calls;
+  }
+};
+
+TEST_F(EnginePMemAllocatorTest, TestBasicAlloc) {
+  uint64_t pmem_size = 128ULL << 20; // 32MB
```
Comment (ZiyanShi) [inline tests/test_pmem_allocator.cpp:167]: 128 MB?

File: tests/test_util.h [hunk 4]
```diff
@@ -4,13 +4,16 @@
...
 #include <cassert>
+#include <chrono>
 #include <iomanip>
 #include <iostream>
 #include <numeric>
 #include <random>
 #include <string>
 #include <vector>
 
+#include "gtest/gtest.h"
```
Comment (ZiyanShi) [inline tests/test_util.h:15]: necessary?

File: tests/tests.cpp [hunk 5]
```diff
@@ -56,6 +68,186 @@ class EngineBasicTest : public testing::Test {
...
+    }
+    return configs;
+  }
+
+  // Set/Get/Delete
+  void GlobalBasicOperator(const std::string &collection, SetOpsFunc SetFunc,
+                           GetOpsFunc GetFunc, DeleteOpsFunc DeleteFunc,
+                           Types type) {
+    // Maybe having create collection for all collection types.
+    if (type == Types::kSorted) {
```
Comment (ZiyanShi) [inline tests/tests.cpp:110]: Don't k an enum. Just Sorted.
k means constant.

**Review by ZiyanShi — CHANGES_REQUESTED** [review]

File: tests/pmem_allocator_bench.cpp [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+#include "kvdk/engine.hpp"
+#include "test_util.h"
+
+using namespace KVDK_NAMESPACE;
+
+#define MIN_ALLOC_SIZE 8
+#define MAX_ALLOC_SIZE 524288 // 512KB
+
+#define NUM_SIZES 10000
+#define NUM_OFFSET 4096
```
Comment (ZiyanShi) [inline tests/pmem_allocator_bench.cpp:28]: Make these parameters benchmark configs, or use constexpr.
Also, give them a more meaningful name.

File: tests/pmem_allocator_bench.cpp [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+
+// Benchmark configs
+DEFINE_string(pmem_path, "/mnt/pmem0/allcator_bench", "Instance path");
+
+DEFINE_uint64(num_thread, 16, "Number of threads");
+
+DEFINE_uint64(iter_num, 10000,
+              "Number of iterating operators(random free and allocate)");
+
+DEFINE_uint64(pmem_size, (std::uint64_t)64 << 30, "PMem total size");
```
Comment (ZiyanShi) [inline tests/pmem_allocator_bench.cpp:38]: (64ULL << 30)
1024ULL

File: tests/pmem_allocator_bench.cpp [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+                          1 / (exponent + 1));
+  }
+
+public:
+  void FixedModePerf(uint32_t num_thread, uint64_t fixed_alloc_size,
+                     int64_t iter_num, int seed, TestAllocator *allocator) {
+    double execute_time = 0;
+    auto TestFixedModePerf = [&](uint64_t id) {
+      std::vector<op_alloc_info> res(iter_num);
+      std::chrono::system_clock::time_point to_begin =
```
Comment (ZiyanShi) [inline tests/pmem_allocator_bench.cpp:74]: auto

File: tests/pmem_allocator_bench.cpp [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+      }
+
+      std::chrono::system_clock::time_point to_end =
+          std::chrono::high_resolution_clock::now();
+      std::chrono::duration<double, std::milli> elapsedTime(to_end - to_begin);
+      execute_time += (elapsedTime.count() / 1000);
+    };
+    LaunchNThreads(num_thread, TestFixedModePerf);
+    std::cout << "Allocated " << std::fixed << std::setprecision(5)
+              << iter_num * fixed_alloc_size
```
Comment (ZiyanShi) [inline tests/pmem_allocator_bench.cpp:90]: Byte?
Second?

File: tests/pmem_allocator_bench.cpp [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+    // Clear all memory to avoid memory leak
+    for (auto record : records) {
+      for (auto r : record) {
+        allocator->wrapped_free(&r);
+      }
+    }
+  }
+
+  AllocatorPerformance(int seed) {
+    std::srand(seed);
```
Comment (ZiyanShi) [inline tests/pmem_allocator_bench.cpp:136]: Consider use std::default_random_engine or std::mt19937

File: tests/test_pmem_allocator.cpp [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+  uint64_t alloc_size = 8;
+
+  // params config
+  std::vector<uint64_t> num_segment_blocks{1024, 2 * 1024, 2 * 1024 * 1024};
+  std::vector<uint32_t> block_sizes{16, 32, 64};
+  std::vector<uint32_t> num_threads = {1, 16};
+
+  for (int i = 0; i < num_segment_blocks.size(); ++i) {
+    for (auto num_thread : num_threads) {
+      thread_manager_.reset(new (std::nothrow) ThreadManager(num_thread));
```
Comment (ZiyanShi) [inline tests/test_pmem_allocator.cpp:58]: Let's make things simple.

File: tests/test_pmem_allocator.cpp [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+          if (space_entry.size == 0)
+            break;
+          records.push_back(space_entry);
+        }
+        ASSERT_EQ(kvpairs, records.size());
+        // TODO: add check pmem usage.
+        delete pmem_alloc;
+      };
+
+      LaunchNThreads(num_thread, TestPmemAlloc);
```
Comment (ZiyanShi) [inline tests/test_pmem_allocator.cpp:91]: Multiple threads with multiple PMEMAllocator?

File: tests/test_pmem_allocator.cpp [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+  uint64_t num_segment_block = 1024;
+  uint64_t block_size = 64;
+  std::vector<uint64_t> alloc_size{8 * 64, 8 * 64, 16 * 64, 32 * 64};
+  thread_manager_.reset(new (std::nothrow) ThreadManager(num_thread));
+  PMEMAllocator *pmem_alloc =
+      PMEMAllocator::NewPMEMAllocator(pmem_path, pmem_size, num_segment_block,
+                                      block_size, num_thread, true, false);
+  ASSERT_NE(pmem_alloc, nullptr);
+
+  /* Allocated pmem status (block nums):
```
Comment (ZiyanShi) [inline tests/test_pmem_allocator.cpp:108]: What about put all these in a vector and then Allocate according to the vector?

File: tests/test_pmem_allocator.cpp [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+  auto TestPmemFree = [&](uint64_t id) {
+    thread_manager_->MaybeInitThread(write_thread);
+    if ((id + 1) % 4 != 0) {
+      pmem_alloc->Free(records[id]);
+      pmem_alloc->BackgroundWork();
+    }
+  };
+
+  // Test merge free memory
+  auto TestPmemFrage = [&](uint64_t id) {
```
Comment (ZiyanShi) [inline tests/test_pmem_allocator.cpp:132]: Frag?

File: tests/test_pmem_allocator.cpp [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+    thread_manager_->MaybeInitThread(write_thread);
+    if ((id + 1 % 4) == 0) {
+      SpaceEntry space_entry = pmem_alloc->Allocate(alloc_size[id % 4]);
+      ASSERT_NE(space_entry.size, 0);
+    }
+  };
+  LaunchNThreads(num_thread, TestPmemFree);
```
Comment (ZiyanShi) [inline tests/test_pmem_allocator.cpp:139]: Is it safe to launch multiple Background threads?

**Review by ZiyanShi — CHANGES_REQUESTED** [review]

File: tests/test_pmem_allocator.cpp (hunk 12 above)
Comment (sherry-1001) [reply tests/test_pmem_allocator.cpp:91]: yes

File: tests/pmem_allocator_bench.cpp (hunk 10 above)
Comment (sherry-1001) [reply tests/pmem_allocator_bench.cpp]: done.

File: tests/pmem_allocator_bench.cpp [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+
+// Allocator Performance Class
+class AllocatorPerformance {
+private:
+  std::vector<std::uint64_t> random_alloc_sizes;
+  std::vector<std::uint64_t> random_offsets;
+
+private:
+  /* Get a random block size with an inverse chi-square distribution.  */
+  static uint64_t get_block_size(uint64_t rand_data) {
```
[the remark this answers is not in the document]
Comment (sherry-1001) [reply tests/pmem_allocator_bench.cpp:52]: Yes, It is actually inverse square distribution. I modified comments.

File: tests/allocator.hpp [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+      t.join();
+    }
+    delete pmem_alloc_;
+  }
+
+private:
+  PMEMAllocator *pmem_alloc_;
+  bool closing_ = false;
+  std::vector<std::thread> background;
+  std::shared_ptr<ThreadManager> thread_manager_;
```
[the remark this answers is not in the document]
Comment (sherry-1001) [reply tests/allocator.hpp:92]: can't use `unique_ptr`.
image

File: tests/CMakeLists.txt [hunk 18]
```diff
@@ -18,3 +17,13 @@ target_include_directories(dbtest_pmem_allocator
     ${PROJECT_SOURCE_DIR}/include
     )
 target_link_libraries(dbtest_pmem_allocator PUBLIC engine gtest gtest_main)
+
+# For pmem allocator bench
+set(BENCH_PMEM_ALLOC pmem_allocator_bench.cpp)
+add_executable(dbbench_pmem_allocator ${BENCH_PMEM_ALLOC})
```
Comment (ZiyanShi) [inline tests/CMakeLists.txt:23]: `add_executable(dbbench_pmem_allocator pmem_allocator_bench.cpp)`

File: tests/allocator.hpp [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+        pmem_path, pmem_size, num_segment_blocks, block_size, num_write_threads,
+        true, false, nullptr);
+    kvdk_assert(pmem_alloc_ != nullptr, "New pmem allocator failed!");
+    background.emplace_back(
+        std::thread(&PMemAllocatorWrapper::BackGround, this));
+  }
+
+  op_alloc_info wrapped_malloc(uint64_t alloc_size) override {
+    op_alloc_info data;
+    data.entry = pmem_alloc_->Allocate(alloc_size);
```
Comment (JiayuZzz) [inline tests/allocator.hpp:61]: return?

File: tests/allocator.hpp [hunk 20]
```diff
@@ -0,0 +1,93 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ * Copyright(c) 2022 Intel Corporation
+ */
+
+#pragma once
+
+#include <stdio.h>
```
Comment (ZiyanShi) [inline tests/allocator.hpp:7]: cstdio

seperate external header from kvdk headers

File: tests/pmem_allocator_bench.cpp [hunk 21]
```diff
@@ -0,0 +1,180 @@
...
+              "Number of iterating operators(random free and allocate)");
+
+DEFINE_uint64(pmem_size, 64ULL << 30, "PMem total size");
+
+DEFINE_uint64(num_segment_blocks, 1024ULL, "PMem num blocks per segment");
+
+DEFINE_uint64(block_size, 16, "PMem block size");
+
+// Allocator Performance Class
+class AllocatorPerformance {
```
Comment (JiayuZzz) [inline tests/pmem_allocator_bench.cpp:45]: Maybe AllocatorBench or something

File: tests/allocator.hpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (ZiyanShi) [reply tests/allocator.hpp:92]: Okay.
But one day we have to use unique_ptr.
Sometimes a write thread will hold the thread manager even when the engine is deleted.

**Review by ZiyanShi — APPROVED** [review]

File: tests/CMakeLists.txt (hunk 18 above)
Comment (sherry-1001) [reply tests/CMakeLists.txt]: done

**Review by JiayuZzz — APPROVED** [review]

**Pull request merged**